src/main/java/com/fakng/fakngagrgtr/parser/google/GoogleParser.java

```diff
@@ -9,57 +15,88 @@
 import org.springframework.web.reactive.function.client.WebClient.ResponseSpec;
 
-import java.io.IOException;
-import java.time.LocalDateTime;
-import java.time.format.DateTimeFormatter;
+import javax.annotation.PostConstruct;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 
 @Component
 public class GoogleParser extends ApiParser {
 
     private static final String GOOGLE_NAME = "Google";
-    private static final DateTimeFormatter DATE_TIME_FORMATTER =
-            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSX");
-    private final Company google;
+    private Company google;
 
-    public GoogleParser(WebClient webClient, CompanyRepository companyRepository, @Value("${url.google}") String url) {
-        super(webClient);
+    public GoogleParser(WebClient webClient, LocationCache locationCache, CompanyRepository companyRepository,
+                        LocationRepository locationRepository, @Value("${url.google}") String url) {
+        super(webClient, locationCache, companyRepository, locationRepository);
         this.url = url;
-        this.google = companyRepository.findByTitle(GOOGLE_NAME).orElse(null);
+    }
+
+    @PostConstruct
+    public void init() {
+        google = companyRepository.findByTitle(GOOGLE_NAME).orElseThrow(() -> new IllegalStateException(String.format("There is no %s company present in DB", GOOGLE_NAME)));
+        google.getLocations().forEach(location -> locationCache.putIfAbsent(location.getCity(), location.getCountry(), location));
     }
 
     @Override
     protected List<Vacancy> getAllVacancies() throws Exception {
-        ResponseDTO firstPage = getPage(1);
+        ResponseDto firstPage = getPage(1);
         int lastPage = firstPage.getCount() / firstPage.getPageSize() + 1;
-        List<Vacancy> vacancies = new ArrayList<>();
-        for (int index = 2; index <= lastPage; index += 1) {
-            List<VacancyDTO> jobs = getPage(index).getJobs();
-            vacancies.addAll(jobs.stream().map(this::createVacancy).toList());
+        List<Vacancy> allVacancies = new ArrayList<>(processPageResponse(firstPage));
+        for (int idx = 2; idx <= lastPage; idx++) {
+            allVacancies.addAll(processPageResponse(getPage(idx)));
         }
-        return vacancies;
+        return allVacancies;
+    }
+
+    private List<Vacancy> processPageResponse(ResponseDto response) {
+        return response.getJobs().stream()
+                .map(this::createVacancy)
+                .toList();
     }
 
-    private Vacancy createVacancy(VacancyDTO dto) {
+    private Vacancy createVacancy(VacancyDto dto) {
         Vacancy vacancy = new Vacancy();
         vacancy.setId(parseVacancyId(dto.getId()));
         vacancy.setTitle(dto.getTitle());
         vacancy.setUrl(dto.getApplyUrl());
         vacancy.setCompany(google);
-        if (dto.getPublishDate() != null) {
-            vacancy.setAddDate(parseLocalDateTime(dto.getPublishDate()));
```

Comment:
Why are you removing this? Vacancy model needs in AddDate field.

Reply:
Add date is rather about entity creation date in DB. Based on those we'll be able to track which vacancies weren't picked up once again => and remove them from DB (cleaning mechanism)

For vacancy's publishing date I'd rather introduce a separate timestamp column in DB for just UI purposes